src/lib/autodiscovery/autodiscovery.py

```diff
@@ -0,0 +1,120 @@
+import os
+import time
+from dataclasses import asdict
+from typing import List
+
+from aiohttp import ClientSession
+
+from lib.autodiscovery.gcp_metrics_descriptor import GCPMetricDescriptor
+from lib.clientsession_provider import init_gcp_client_session
+from lib.configuration import config
+from lib.context import LoggingContext
+from lib.credentials import create_token
+from lib.metrics import GCPService, Metric
+
+logging_context = LoggingContext("AUTODISCOVERY")
+
+
+discovered_resource_type = os.environ.get("AUTODISCOVERY_RESOURCE_TYPE", "gce_instance")
+
+
+async def get_metric_descriptors(
+    gcp_session: ClientSession, token: str
+) -> List[GCPMetricDescriptor]:
+    project_id = config.project_id()
+    headers = {"Accept": "application/json", "Authorization": f"Bearer {token}"}
+    url = f"https://monitoring.googleapis.com/v3/projects/{project_id}/metricDescriptors"
+    params = {}
+
+    response = await gcp_session.request("GET", url=url, headers=headers, params=params)
+    response = await response.json()
+
+    discovered_metrics_descriptors = []
+
+    while True:
+        discovered_metrics_descriptors.extend(
+            [GCPMetricDescriptor(**descriptor) for descriptor in response.get("metricDescriptors",[])]
+        )
+
+        page_token = response.get("nextPageToken", "")
+        params["pageToken"] = page_token
+        if page_token == "":
+            break
+
+    discovered_metrics_descriptors = list(
+        filter(
+            lambda descriptor: descriptor.gcpOptions.valueType.upper() != "STRING",
+            discovered_metrics_descriptors,
+        )
+    )
+    discovered_metrics_descriptors = list(
+        filter(
+            lambda descriptor: discovered_resource_type in descriptor.monitored_resources_types,
+            discovered_metrics_descriptors,
+        )
+    )
+    return discovered_metrics_descriptors
+
+
+async def run_autodiscovery(
+    gcp_services_list: List[GCPService], gcp_session: ClientSession, token: str
+):
+    start_time = time.time()
+    logging_context.log("Adding metrics using autodiscovery")
+
+    bucket_gcp_services = list(
+        filter(lambda x: discovered_resource_type in x.name, gcp_services_list)
+    )
+
+    existing_metric_list = []
+    for service in bucket_gcp_services:
+        existing_metric_list.extend(service.metrics)
+
+    discovered_metric_descriptors = await get_metric_descriptors(gcp_session, token)
+
+    existing_metric_names = [
+        existing_metric.google_metric for existing_metric in existing_metric_list
+    ]
+    missing_metrics_list = []
+
+    for descriptor in discovered_metric_descriptors:
+        if descriptor.value not in existing_metric_names:
+            missing_metrics_list.append(Metric(**asdict(descriptor)))
+    logging_context.log(f"In Extension have: {len(existing_metric_list)} metrics")
```

Comment:
E.g.: "In the extension there are ... metrics", something like that. Just a comment about redaction.
Feel free to change it later if you want to merge, not urgent.

Reply:
PS: Sorry I didn't write on the previous comment, I took it as a suggestions, not literal 😄